## **🧭 Task 1 – Fundamental Data Understanding (Indian Air Pollution Data)**

### **🎯 Objective :**

The purpose of this task is to help you explore and understand the structure and quality of the Indian Air Pollution Dataset provided for your assessment.
You will use Pandas to:

* Combine multiple CSV files into a single dataset, and

* Perform fundamental data understanding (data inspection, summary statistics, and missing value analysis).

## **Mounting the drive**

In [ ]:
# Load the Drive helper and mount
from google.colab import drive

# This will prompt for authorization.
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
%cd '/content/drive/MyDrive/Assessment Data-20251111'
# please change the path according to the location of your data

/content/drive/MyDrive/Assessment Data-20251111


In [ ]:
%ls # it shows all the content of your folder if done properly

Ahmedabad_data.csv     Coimbatore_data.csv  Kolkata_data.csv
Aizawl_data.csv        Delhi_data.csv       Lucknow_data.csv
Amaravati_data.csv     Ernakulam_data.csv   Mumbai_data.csv
Amritsar_data.csv      Gurugram_data.csv    Patna_data.csv
Bengaluru_data.csv     Guwahati_data.csv    Shillong_data.csv
Bhopal_data.csv        Hyderabad_data.csv   Talcher_data.csv
Brajrajnagar_data.csv  Jaipur_data.csv      Thiruvananthapuram_data.csv
Chandigarh_data.csv    Jorapokhar_data.csv  Visakhapatnam_data.csv
Chennai_data.csv       Kochi_data.csv


# **🧩 What You Need to Do:**

**Step 1: Import Libraries :**

Start by importing the necessary Python libraries:

In [ ]:
import pandas as pd
import os

### **🧩 Merging of the csv files:**



### **Combining all the csv files in the drive path: You can use either of the merging options:**  
---

**🧩 OPTION 1 :**

In [ ]:
import pandas as pd
import glob
import os
from google.colab import drive

# Ensure Google Drive is mounted before proceeding
# This will prompt for authorization if not already mounted
drive.mount('/content/drive')

drive_path = '/content/drive/MyDrive/Assessment Data-20251111' # Corrected path

# Now glob.glob will search directly in the specified drive_path
city_files = glob.glob(os.path.join(drive_path, "*.csv"))

all_cities_data = []

if not city_files:
    print(f"No CSV files found in {drive_path}. Please check the path and ensure files exist.")
elif city_files:
    for file_name in city_files:
        city_df = pd.read_csv(file_name)
        all_cities_data.append(city_df)
        print(f"Loaded: {file_name}")

    if all_cities_data:
        combined_data = pd.concat(all_cities_data, ignore_index=True)
        output_file_path = os.path.join(drive_path, "all_cities_combined.csv")
        combined_data.to_csv(output_file_path, index=False)

        print(f"SUCCESS: Combined {len(city_files)} city files into one file with {len(combined_data)} total rows")
        print(f"The combined file is saved as: {output_file_path}")
    else:
        print("No data to combine after loading files.")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Lucknow_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Thiruvananthapuram_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Bengaluru_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Hyderabad_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Chennai_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Jorapokhar_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Aizawl_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Amritsar_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Ernakulam_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Coimbatore_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Brajrajnagar_data.csv
Loaded: /content/drive/MyDrive/Assessmen

In [ ]:
df = pd.read_csv(output_file_path)
df

City        Date  PM2.5    PM10    NO    NO2    NOx    NH3     CO  \
0      Lucknow  01/01/2015    NaN     NaN  2.11  13.46   4.57    NaN  12.15   
1      Lucknow  02/01/2015    NaN     NaN  2.37  10.76   3.80    NaN  14.00   
2      Lucknow  03/01/2015    NaN     NaN  2.27  12.43   4.22    NaN  12.78   
3      Lucknow  04/01/2015    NaN     NaN  2.24  12.46   4.22    NaN  11.85   
4      Lucknow  05/01/2015    NaN     NaN  2.32  12.27   4.22    NaN  13.34   
...        ...         ...    ...     ...   ...    ...    ...    ...    ...   
29526   Jaipur  27/06/2020  28.33   76.66  6.15  14.72  18.14  21.82   0.59   
29527   Jaipur  28/06/2020  23.76   67.44  5.60  14.05  17.20  20.41   0.50   
29528   Jaipur  29/06/2020  30.58   77.30  6.21  16.63  19.50  22.70   0.55   
29529   Jaipur  30/06/2020  30.03   64.26  5.48  13.77  16.94  19.79   0.67   
29530   Jaipur  01/07/2020  58.79  103.68  5.19  10.88  14.16  17.41   0.54   

          SO2     O3  Benzene  Toluene  Xylene   AQI    AQI_Bucket  
0      169.57  25.92     1.35     3.93     NaN   NaN           NaN  
1      169.46  21.66     1.97     5.74     NaN   NaN           NaN  
2      169.77  22.40     1.83     4.76     NaN   NaN           NaN  
3      169.56  21.41     2.27     4.77     NaN   NaN           NaN  
4      169.48  20.60     2.49     4.78     NaN   NaN           NaN  
...       ...    ...      ...      ...     ...   ...           ...  
29526    9.96  39.12     0.23     0.59     NaN  89.0  Satisfactory  
29527    8.93  34.40     0.16     0.39     NaN  75.0  Satisfactory  
29528   10.59  39.21     0.22     0.47     NaN  69.0  Satisfactory  
29529   10.36  50.71     0.48     1.48     NaN  79.0  Satisfactory  
29530    5.19  53.68     0.22     0.46     NaN  70.0  Satisfactory  

[29531 rows x 16 columns]

**🧩 OPTION 2 :**


The process involves iterating through all files in a specified directory (drive_path) to identify those with a .csv extension. Each identified CSV file is then read into a Pandas DataFrame, which is subsequently appended to a list (dataframes). Once all the CSV files are processed, the individual DataFrames in the list are combined into a single consolidated DataFrame (all_data) using the pd.concat() function. This consolidation ensures that the combined data is reindexed, creating a unified dataset for further analysis.

In [ ]:
drive_path = '/content/drive/MyDrive/Assessment Data-20251111' # Corrected path

In [ ]:
import pandas as pd
import glob
import os
from google.colab import drive

# Ensure Google Drive is mounted before proceeding
# This will prompt for authorization if not already mounted
drive.mount('/content/drive')

drive_path = '/content/drive/MyDrive/Assessment Data-20251111' # Corrected path

# Now glob.glob will search directly in the specified drive_path
city_files = glob.glob(os.path.join(drive_path, "*.csv"))

all_cities_data = []

if not city_files:
    print(f"No CSV files found in {drive_path}. Please check the path and ensure files exist.")
elif city_files:
    for file_name in city_files:
        city_df = pd.read_csv(file_name)
        all_cities_data.append(city_df)
        print(f"Loaded: {file_name}")

    if all_cities_data:
        combined_data = pd.concat(all_cities_data, ignore_index=True)
        output_file_path = os.path.join(drive_path, "all_cities_combined.csv")
        combined_data.to_csv(output_file_path, index=False)

        print(f"SUCCESS: Combined {len(city_files)} city files into one file with {len(combined_data)} total rows")
        print(f"The combined file is saved as: {output_file_path}")
    else:
        print("No data to combine after loading files.")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Lucknow_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Thiruvananthapuram_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Bengaluru_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Hyderabad_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Chennai_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Jorapokhar_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Aizawl_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Amritsar_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Ernakulam_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Coimbatore_data.csv
Loaded: /content/drive/MyDrive/Assessment Data-20251111/Brajrajnagar_data.csv
Loaded: /content/drive/MyDrive/Assessmen

In [ ]:
dataframes = []
for filename in os.listdir(drive_path):
    if filename.endswith('.csv'):  # Check if the file is a CSV file
        file_path = os.path.join(drive_path, filename)
        df = pd.read_csv(file_path)  # Read the CSV file into a DataFrame
        dataframes.append(df)  # Add the DataFrame to the list

In [ ]:
df = pd.concat(dataframes, ignore_index=True)

In [ ]:
print('Average levels of key pollutants:')
print(df[['PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2', 'O3', 'Benzene', 'Toluene', 'Xylene']].mean())

Average levels of key pollutants:
PM2.5       67.450578
PM10       118.127103
NO          17.574730
NO2         28.560659
NOx         32.309123
NH3         23.483476
CO           2.248598
SO2         14.531977
O3          34.491430
Benzene      3.280840
Toluene      8.700972
Xylene       3.070128
dtype: float64


## **🧩 Perform Fundamental Data Understanding :**

**Once you have the merged dataset, explore and understand its structure.**

In [ ]:
df.head(10)

City        Date  PM2.5  PM10    NO    NO2   NOx  NH3     CO     SO2  \
0  Lucknow  01/01/2015    NaN   NaN  2.11  13.46  4.57  NaN  12.15  169.57   
1  Lucknow  02/01/2015    NaN   NaN  2.37  10.76  3.80  NaN  14.00  169.46   
2  Lucknow  03/01/2015    NaN   NaN  2.27  12.43  4.22  NaN  12.78  169.77   
3  Lucknow  04/01/2015    NaN   NaN  2.24  12.46  4.22  NaN  11.85  169.56   
4  Lucknow  05/01/2015    NaN   NaN  2.32  12.27  4.22  NaN  13.34  169.48   
5  Lucknow  06/01/2015    NaN   NaN  2.33  12.91  4.30  NaN  14.21  178.63   
6  Lucknow  07/01/2015    NaN   NaN  2.68  17.12  4.64  NaN  11.24  169.00   
7  Lucknow  08/01/2015    NaN   NaN  4.19  24.45  3.97  NaN  12.68  169.00   
8  Lucknow  09/01/2015    NaN   NaN  4.27  24.74  3.13  NaN  13.99  182.39   
9  Lucknow  10/01/2015    NaN   NaN  4.02  20.67  4.05  NaN  14.06  174.58   

      O3  Benzene  Toluene  Xylene  AQI AQI_Bucket  
0  25.92     1.35     3.93     NaN  NaN        NaN  
1  21.66     1.97     5.74     NaN  NaN        NaN  
2  22.40     1.83     4.76     NaN  NaN        NaN  
3  21.41     2.27     4.77     NaN  NaN        NaN  
4  20.60     2.49     4.78     NaN  NaN        NaN  
5  21.49     2.93     4.60     NaN  NaN        NaN  
6  23.67     2.92     3.82     NaN  NaN        NaN  
7  19.69     3.01     5.38     NaN  NaN        NaN  
8  21.73     2.52     4.78     NaN  NaN        NaN  
9  27.36     2.55     4.98     NaN  NaN        NaN

In [ ]:
df.tail(10)

City        Date  PM2.5    PM10    NO    NO2    NOx    NH3    CO  \
29521  Jaipur  22/06/2020  34.32   67.36  5.51  16.12  18.65  25.60  0.68   
29522  Jaipur  23/06/2020  28.29   56.81  5.39  14.50  17.77  20.99  0.64   
29523  Jaipur  24/06/2020  24.06   51.10  5.83  14.37  17.57  21.54  0.70   
29524  Jaipur  25/06/2020  64.47  202.53  6.49  16.88  20.25  24.40  0.67   
29525  Jaipur  26/06/2020  35.81  118.34  6.61  18.81  21.85  25.43  0.69   
29526  Jaipur  27/06/2020  28.33   76.66  6.15  14.72  18.14  21.82  0.59   
29527  Jaipur  28/06/2020  23.76   67.44  5.60  14.05  17.20  20.41  0.50   
29528  Jaipur  29/06/2020  30.58   77.30  6.21  16.63  19.50  22.70  0.55   
29529  Jaipur  30/06/2020  30.03   64.26  5.48  13.77  16.94  19.79  0.67   
29530  Jaipur  01/07/2020  58.79  103.68  5.19  10.88  14.16  17.41  0.54   

         SO2     O3  Benzene  Toluene  Xylene    AQI    AQI_Bucket  
29521  10.04  53.53     0.47     0.79     NaN   80.0  Satisfactory  
29522   9.77  64.22     0.36     0.63     NaN   80.0  Satisfactory  
29523   9.68  46.24     0.67     1.21     NaN   72.0  Satisfactory  
29524   8.76  45.26     0.59     1.58     NaN   76.0  Satisfactory  
29525  10.34  39.55     0.50     1.71     NaN  172.0      Moderate  
29526   9.96  39.12     0.23     0.59     NaN   89.0  Satisfactory  
29527   8.93  34.40     0.16     0.39     NaN   75.0  Satisfactory  
29528  10.59  39.21     0.22     0.47     NaN   69.0  Satisfactory  
29529  10.36  50.71     0.48     1.48     NaN   79.0  Satisfactory  
29530   5.19  53.68     0.22     0.46     NaN   70.0  Satisfactory

In [ ]:
print(f'No of Rows: {df.shape[0]}, No of Columns: {df.shape[1]}')

No of Rows: 88593, No of Columns: 16


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88593 entries, 0 to 88592
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   City        88593 non-null  object 
 1   Date        88593 non-null  object 
 2   PM2.5       74799 non-null  float64
 3   PM10        55173 non-null  float64
 4   NO          77847 non-null  float64
 5   NO2         77838 non-null  float64
 6   NOx         76038 non-null  float64
 7   NH3         57609 non-null  float64
 8   CO          82416 non-null  float64
 9   SO2         77031 non-null  float64
 10  O3          76527 non-null  float64
 11  Benzene     71724 non-null  float64
 12  Toluene     64470 non-null  float64
 13  Xylene      34266 non-null  float64
 14  AQI         74550 non-null  float64
 15  AQI_Bucket  74550 non-null  object 
dtypes: float64(13), object(3)
memory usage: 10.8+ MB


In [ ]:
df.dtypes

City           object
Date           object
PM2.5         float64
PM10          float64
NO            float64
NO2           float64
NOx           float64
NH3           float64
CO            float64
SO2           float64
O3            float64
Benzene       float64
Toluene       float64
Xylene        float64
AQI           float64
AQI_Bucket     object
dtype: object

In [ ]:
df.columns

Index(['City', 'Date', 'PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2',
       'O3', 'Benzene', 'Toluene', 'Xylene', 'AQI', 'AQI_Bucket'],
      dtype='object')

### **Total number of stations in the dataset:**

In [ ]:
unique_cities = df['City'].nunique()
print(f'Total number of unique cities in the dataset: {unique_cities}')
df['City'].value_counts()

Total number of unique cities in the dataset: 26


City
Lucknow               6027
Bengaluru             6027
Chennai               6027
Ahmedabad             6027
Delhi                 6027
Mumbai                6027
Hyderabad             6018
Patna                 5574
Gurugram              5037
Visakhapatnam         4386
Amritsar              3663
Jorapokhar            3507
Jaipur                3342
Thiruvananthapuram    3336
Amaravati             2853
Brajrajnagar          2814
Talcher               2775
Kolkata               2442
Guwahati              1506
Coimbatore            1158
Shillong               930
Chandigarh             912
Bhopal                 867
Ernakulam              486
Kochi                  486
Aizawl                 339
Name: count, dtype: int64

### **Displaying the percentage of missing value**

In [ ]:
def missing_values_table(df1):
    # Total missing values
    mis_val = df1.isnull().sum()

    # Percentage of missing values
    mis_val_percent = 100 * mis_val / len(df1)

    mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)

    mis_val_table = mis_val_table.rename(columns={0: 'Missing Values', 1: '% of Total Values'})

    # Sort the table by percentage of missing descending
    mis_val_table = mis_val_table.sort_values('% of Total Values', ascending=False)

    return mis_val_table

missing_values = missing_values_table(df)
display(missing_values.style.background_gradient(cmap='Blues'))

# **📊 Analysis Questions to Answer :**

**Include short written answers (3–5 sentences) to these:**

* How many rows and columns are in your merged dataset?
29531 rows × 16 columns
* Which pollutants are included, and which have the most missing data?
Xylene, PM10, NH3,Toluene,Benzene,AQI_Bucket,PM2.5,NOx,O3,SO2,NO2,NO,CO

* How many unique cities or stations are there?
26 Chennai,Lucknow,Ahmedabad,Delhi,Mumbai,Bengaluru,Hyderabad,Patna,Gurugram,Visakhapatnam,Amritsar,Jorapokhar,Jaipur,Thiruvananthapuram,Amaravati,Brajrajnagar,Talcher,Kolkata,Guwahati,Coimbatore ,Shillong,Chandigarh, Bhopal, Ernakulam ,Kochi,Aizawl

* What are the average levels of key pollutants?
Average levels of key pollutants:
PM2.5       67.450578
PM10       118.127103
NO          17.574730
NO2         28.560659
NOx         32.309123
NH3         23.483476
CO           2.248598
SO2         14.531977
O3          34.491430
Benzene      3.280840
Toluene      8.700972
Xylene       3.070128

* Are there any immediate data quality issues (e.g., missing or inconsistent values)?
yes